### PR TITLE
Converting HttpClientBuilders to interfaces

### PR DESCRIPTION
**Pull request by chemicL**
Motivation:

The hierarchy of client builders has recently been simplified. It is now
possible to simplify the structure even more by avoiding abstract
classes with intermediate implementations in the base abstract class for
particular case and using straightforward interfaces with default
implementations in the `http-netty` package.

Modifications:

- `MultiAddressHttpClientBuilder`, `PartitionedHttpClientBuilder`, and
  `SingleAddressHttpClientBuilder` are now interfaces,
- `DefaultMultiAddressUrlHttpClientBuilder`,
  `DefaultPartitionedHttpClientBuilder`, and
  `DefaultSingleAddressHttpClientBuilder` implement the above interfaces
  and the entire functionality is encapsulated within those final
  classes.

Result:

Cleaner API with interfaces and no intermediate abstract classes with
implementation scattered between the base and final classes.

### CODE REVIEW

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java:75]: Same

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:169]: Same

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:75]: Shifted indentation

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:246]: Shifted indentation

**Review by idelpivnitskiy — COMMENTED** [review]
LGTM after comments addressed.
If you don't mind, can we hold this PR as well until after #1792? Redirect should also be cherry-picked into 0.41. That will help to avoid conflicts in `MultiAddressHttpClientBuilder`.

**Review by bondolo — APPROVED** [review]
Please mark conversations as resolved once the PR incorporates the suggested changes. Leaving the conversations unresolved creates the impression that there are still additional changes to be made.  (Which can lead to the impression that review should be deferred until those changes are available).

**Comment by idelpivnitskiy** [comment]
Feel free to merge this PR after #1844 to close this task, don't wait for #1792. The conflicts will be minimal.

**Pull request merged**
